### PR TITLE
feat(tokenizer):add support for open source llm tokenizers

### LINKED ISSUE #1666: [Issue]: Autogen count tokens badly when using served Open Source Models
### Describe the issue

I've been using various open source models (using vllm to serve them: Orca-2-7b, Mixtral_11Bx2_MoE_19B, and SOLAR-10.7B-Instruct-v1.0) to implement RAG using Qdrant. These models have been included in my OAI_CONFIG_LIST and appear to function well using autogen. However, I've encountered an issue when introducing larger prompts, the number of tokens are always calculated badly.

Upon inspection of the source code, I've deduced that the token count is calculated using Tiktoken, whereas the open source models employ a different calculation method. This discrepancy seems to be causing issues when the token count surpasses the maximum token length of the model, resulting in a process halt.
I'll come with a PR asap.

### Steps to reproduce

Serve an open source model using vllm, just use it with a big prompt.

### Additional Information

### PR DESCRIPTION
**Pull request by Halpph**
Hello everyone, I saw the Contribute guide but for some reason the tests would always fail on `_____________________________________ ERROR collecting test/test_function_utils.py _____________________________________
test/test_function_utils.py:288: in <module>
    class Currency(BaseModel):
pydantic/main.py:197: in pydantic.main.ModelMetaclass.__new__
    ???
pydantic/fields.py:497: in pydantic.fields.ModelField.infer
    ???
pydantic/fields.py:469: in pydantic.fields.ModelField._get_field_info
    ???
E   ValueError: `Field` default cannot be set in `Annotated` for 'amount'`

I tried but I'm very busy and I seem to not manage to make it work for now, I hope you can take a look and I'll try to run it again.

## Why are these changes needed?

This PR solves the following issue [https://github.com/microsoft/autogen/issues/1666](url)
Basically while serving open source llm models we were always tokenizing using cl100k_base, but now we support the native way of each model by specifying it in the OAI_CONFIG_LIST

## Related issue number

Closes #1666 

## Checks
Sadly I didn't manage to run checks because of the error mentioned above

### CODE REVIEW

**Comment by Halpph** [comment]
@microsoft-github-policy-service agree

**Comment by Halpph** [comment]
@olgavrou @AaronWard @kevin666aa @SDcodehub

Hello everyone, this is a workaround that I implemented for now in order to be able to continue with my work, but I'd like to discuss with you of a proper implementation before starting the refactoring.
For instance we could implement this using some kind of strategy pattern so that the desired strategy would be responsible for fetching the right tokenizer for each case. What do you think?

**Comment by ekzhu** [comment]
@Halpph is this still the workaround you are using?

**Comment by Halpph** [comment]
I'm not actively using it at the moment, but yes.

**Comment by rysweet** [comment]
This PR is against AutoGen 0.2. AutoGen 0.2 has been moved to the 0.2 branch. Please rebase your PR on the 0.2 branch or update it to work with the new AutoGen 0.4 that is now in main.

**Comment by rysweet** [comment]
@Halpph If you can update to resolve conflicts and see if we can get CI to pass we can look at bringing this forward

**Comment by rysweet** [comment]
closing as stale, please reopen if you would like to bring it up to date.
